### PR TITLE
Add optional retries to Auto-Approve Apply and Destroy Commands

### PR DESCRIPTION
**Pull request by benwart**
Changes Made:
- Added `retries` and `retry_delay` parameters to Apply and Destroy Commands
- Pipe `retries` and `retry_delay` from Apply and Destroy Jobs into appropriate Commands
- Added check for `GITHUB_*` envvars in github.py and fail fast with simple error message rather than exception stack trace

### CODE REVIEW

**Comment by danielflookovo** [comment]
The absence of `GITHUB_` variables should not cause the job to fail. There should be better messaging than a stacktrace though.

Can you explain the use case for `retry` and `retry_delay`.

**Comment by benwart** [comment]
The use case for retry is that during apply and destroy actions terraform will sometimes fail (mostly for timeouts).  On retry they will succeed.  The `retries` parameter is the count of retries to attempt before failing and the `retry_delay` parameter is the sleep time between retries.  It is only for apply and destroy actions and should maintain the behavior that was originally in place with the final attempt setting the exit code used to determine pass/fail.

File: terraform/destroy.sh [hunk 1]
```diff
@@ -1 +1,17 @@
...
+
+for ((i=0; i <= $RETRIES)); do
+  TF_EXIT=terraform destroy -input=false -no-color -auto-approve $PLAN_ARGS "$module_path"
+
+  if [[ $TF_EXIT -eq 0 ]]; then
+    break
+  fi
+
+  if [[ $i -nq $RETRIES ]]; then
+    sleep $RETRY_DELAY
```
[the remark this answers is not in the document]
Comment (benwart) [reply terraform/destroy.sh]: Progress for the sleep of the number of attempts?

File: terraform/github.py [hunk 2]
```diff
@@ -7,8 +7,13 @@
 
 import requests
 
-github_username = os.environ['GITHUB_USERNAME']
-github_token = os.environ['GITHUB_TOKEN']
+github_username = os.environ.get('GITHUB_USERNAME')
+github_token = os.environ.get('GITHUB_TOKEN')
+
+if github_username == None or github_token == None:
+    print(f'Missing GITHUB_USERNAME or GITHUB_TOKEN in Environment')
```
[the remark this answers is not in the document]
Comment (benwart) [reply terraform/github.py]: Why is this redundant?  I'm not sure I see what you are referring to.

**Comment by benwart** [comment]
Thanks for getting the feedback.  I did not test this before submitting initially but I have tested this now and found a couple of items that are fixed and included in the PR.  I have also added comments for the specific review items that I'm not clear on.

File: terraform/github.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (takac) [reply terraform/github.py]: it errors out and gives a fairly clear `KeyError: GITHUB_TOKEN` without this change. I appreciate that your changes gives a more context to the issue.

File: terraform/apply.sh [hunk 3]
```diff
@@ -40,35 +61,34 @@ fi
...
+    fi
+
     if ! python3 /tmp/github.py get >approved-plan.txt; then
         echo "Approved plan not found"
         exit 1
     fi
 
     if python3 /tmp/cmp.py plan.txt approved-plan.txt; then
-        apply
+        exit $(apply)
```
Comment (takac) [inline terraform/apply.sh]: This doesn't pass through the exit code

**Comment by bamarch** [comment]
@benwart is there a chance you'll come back to this and update the conflicts? :D

**Comment by benwart** [comment]
@bamarch yes, I fixed the conflicts.

**Pull request closed without merging**